File: packages/sdks/server-sdk/src/services/GameWorldService.ts
/**
 * Game World Service
 * Handles game world list and detail retrieval
 * Uses per-environment API pattern: GET /api/v1/server/:env/game-worlds
 * Extends BaseEnvironmentService for common fetch/caching logic
 */

import { ApiClient } from '../client/ApiClient';
import { Logger } from '../utils/logger';
import { EnvironmentResolver } from '../utils/EnvironmentResolver';
import { GameWorld, GameWorldListResponse } from '../types/api';
import { BaseEnvironmentService } from './BaseEnvironmentService';

export class GameWorldService extends BaseEnvironmentService<GameWorld, GameWorldListResponse, number> {
  constructor(apiClient: ApiClient, logger: Logger, envResolver: EnvironmentResolver) {
    super(apiClient, logger, envResolver);
  }

  // ==================== Abstract Method Implementations ====================

  protected getEndpoint(environment: string): string {
    return `/api/v1/server/${encodeURIComponent(environment)}/game-worlds`;
  }

  protected extractItems(response: GameWorldListResponse): GameWorld[] {
    return response.worlds;
  }

  protected getServiceName(): string {
    return 'game worlds';
  }

  protected getItemId(item: GameWorld): number {
    return item.id;
  }

  // ==================== Override for Custom Sorting ====================

  /**
   * Get game worlds for a specific environment
   * Overridden to add sorting by displayOrder
   */
  async listByEnvironment(environment: string): Promise<GameWorld[]> {
    const endpoint = this.getEndpoint(environment);

    this.logger.debug('Fetching game worlds', { environment });

    const response = await this.apiClient.get<GameWorldListResponse>(endpoint);

    if (!response.success || !response.data) {
      throw new Error(response.error?.message || 'Failed to fetch game worlds');
    }

    const worlds = this.extractItems(response.data);
    // Sort by displayOrder (ascending)
    const sortedWorlds = worlds.sort((a, b) => (a.displayOrder || 0) - (b.displayOrder || 0));
    this.cachedByEnv.set(environment, sortedWorlds);

    this.logger.info('Game worlds fetched', { count: sortedWorlds.length, environment });

    return sortedWorlds;
  }

  // ==================== Domain-specific Methods ====================

  /**
   * Get game world by ID
   * GET /api/v1/server/:env/game-worlds/:id
   * @param id Game world ID
   * @param environment Environment name (required)
   */
  async getById(id: number, environment: string): Promise<GameWorld> {
    this.logger.debug('Fetching game world by ID', { id, environment });

    const response = await this.apiClient.get<GameWorld>(`/api/v1/server/${encodeURIComponent(environment)}/game-worlds/${id}`);

    if (!response.success || !response.data) {
      throw new Error(response.error?.message || 'Failed to fetch game world');
    }

    this.logger.info('Game world fetched', { id, worldId: response.data.worldId });

    return response.data;
  }

  /**
   * Get game world by worldId
   * GET /api/v1/server/:env/game-worlds/world/:worldId
   * @param worldId World ID
   * @param environment Environment name (required)
   */
  async getByWorldId(worldId: string, environment: string): Promise<GameWorld> {
    this.logger.debug('Fetching game world by worldId', { worldId, environment });

    const response = await this.apiClient.get<GameWorld>(
      `/api/v1/server/${encodeURIComponent(environment)}/game-worlds/world/${encodeURIComponent(worldId)}`
    );

    if (!response.success || !response.data) {
      throw new Error(response.error?.message || 'Failed to fetch game world');
    }

    this.logger.info('Game world fetched', { worldId, id: response.data.id });

    return response.data;
  }

  /**
   * Update a single game world in cache (immutable)
   * If isVisible is false, removes the world from cache (no API call needed)
   * If isVisible is true but not in cache, fetches and adds it to cache
   * If isVisible is true and in cache, fetches and updates it
   * @param id Game world ID
   * @param environment Environment name (required)
   * @param isVisible Optional visibility status
   */
  async updateSingleWorld(id: number, environment: string, isVisible?: boolean | number): Promise<void> {
    try {
      this.logger.debug('Updating single game world in cache', { id, environment, isVisible });

      // If isVisible is explicitly false (0 or false), just remove from cache
      if (isVisible === false || isVisible === 0) {
        this.logger.info('Game world isVisible=false, removing from cache', { id, environment });
        this.removeFromCache(id, environment);
        return;
      }

      // Otherwise, fetch from API and add/update
      // Add small delay to ensure backend transaction is committed
      await new Promise(resolve => setTimeout(resolve, 100));

      // Fetch the updated world from backend
      const updatedWorld = await this.getById(id, environment);

      // Get current worlds for this environment
      const currentWorlds = this.cachedByEnv.get(environment) || [];
      const existsInCache = currentWorlds.some(world => world.id === id);

      if (existsInCache) {
        // Immutable update: update existing world
        const newWorlds = currentWorlds.map(world => world.id === id ? updatedWorld : world);
        this.cachedByEnv.set(environment, newWorlds);
        this.logger.debug('Single game world updated in cache', { id, environment });
      } else {
        // World not in cache but found in backend
        // Add it to cache and re-sort by displayOrder
        const newWorlds = [...currentWorlds, updatedWorld];
        newWorlds.sort((a, b) => (a.displayOrder || 0) - (b.displayOrder || 0));
        this.cachedByEnv.set(environment, newWorlds);
        this.logger.debug('Single game world added to cache', { id, environment });
      }
    } catch (error: any) {
      this.logger.error('Failed to update single game world in cache', {
        id,
        environment,
        error: error.message,
      });
      // If update fails, fall back to full refresh
      await this.refreshByEnvironment(environment);
    }
  }

  /**
   * Check if a world is in maintenance based on flag and time window
   * @param worldId World ID
   * @param environment Environment name (required)
   */
  isWorldMaintenanceActive(worldId: string, environment: string): boolean {
    const worlds = this.getCached(environment);
    const world = worlds.find((w) => w.worldId === worldId);
    if (!world || !world.isMaintenance) {
      return false;
    }

    const now = new Date();

    // Check if maintenance has not started yet
    if (world.maintenanceStartDate) {
      const startDate = new Date(world.maintenanceStartDate);
      if (!Number.isNaN(startDate.getTime()) && now < startDate) {
        return false;
      }
    }

    // Check if maintenance has already ended
    if (world.maintenanceEndDate) {
      const endDate = new Date(world.maintenanceEndDate);
      if (!Number.isNaN(endDate.getTime()) && now > endDate) {
        return false;
      }
    }

    return true;
  }

  /**
   * Get world by worldId from cache
   * @param worldId World ID
   * @param environment Environment name (required)
   */
  getWorldByWorldId(worldId: string, environment: string): GameWorld | null {
    const worlds = this.getCached(environment);
    return worlds.find((w) => w.worldId === worldId) || null;
  }

  /**
   * Get maintenance message for a world with language support
   * Falls back to default message if language not found
   * @param worldId World ID
   * @param environment Environment name (required)
   * @param lang Language code (default: 'en')
   */
  getWorldMaintenanceMessage(worldId: string, environment: string, lang: 'ko' | 'en' | 'zh' = 'en'): string | null {
    const worlds = this.getCached(environment);
    const world = worlds.find((w) => w.worldId === worldId);
    if (!world || !this.isWorldMaintenanceActive(worldId, environment)) {
      return null;
    }

    // Try to find localized message
    if (world.maintenanceLocales && world.maintenanceLocales.length > 0) {
      const locale = world.maintenanceLocales.find((l: any) => l.lang === lang);
      if (locale) {
        return locale.message;
      }
    }

    // Fallback to default message
    return world.maintenanceMessage || null;
  }
}

File: packages/sdks/server-sdks/gatrix-dotnet-server-sdk/src/Gatrix.Server.Sdk/Services/GameWorldService.cs
using Gatrix.Server.Sdk.Cache;
using Gatrix.Server.Sdk.Client;
using Gatrix.Server.Sdk.Models;
using Microsoft.Extensions.Logging;

namespace Gatrix.Server.Sdk.Services;

public interface IGameWorldService
{
    Task InitializeAsync(string environmentId, CancellationToken ct = default);
    Task<List<GameWorld>> FetchAsync(string environmentId, CancellationToken ct = default);
    List<GameWorld> GetCached(string environmentId);
    List<GameWorld> GetAll(string environmentId);
    GameWorld? GetByWorldId(string worldId, string environmentId);
    bool IsWorldMaintenanceActive(string worldId, string environmentId);
    string? GetWorldMaintenanceMessage(string worldId, string environmentId, string lang = "en");

    /// <summary>
    /// Update a single game world in cache.
    /// If isVisible=false, removes from cache (no API call).
    /// If isVisible=true, fetches single item from API and upserts.
    /// Falls back to full refresh on failure.
    /// </summary>
    Task UpdateSingleWorldAsync(int id, string environmentId, bool? isVisible = null, CancellationToken ct = default);

    /// <summary>Remove a game world from cache by id.</summary>
    void RemoveWorld(int id, string environmentId);
}

public class GameWorldService : BaseEnvironmentService<GameWorld, GameWorldListResponse>, IGameWorldService
{
    public GameWorldService(GatrixApiClient apiClient, ILogger<GameWorldService> logger, ICacheStorageProvider? storage = null)
        : base(apiClient, logger, storage) { }

    protected override string ServiceName => "GameWorld";
    protected override string GetEndpoint(string environmentId) =>
        $"/api/v1/server/game-worlds";
    protected override List<GameWorld> ExtractItems(GameWorldListResponse response) =>
        response.Worlds.OrderBy(w => w.DisplayOrder).ToList();
    protected override object GetItemId(GameWorld item) => item.Id;

    public Task<List<GameWorld>> FetchAsync(string environmentId, CancellationToken ct = default) =>
        FetchByEnvironmentAsync(environmentId, ct);

    public List<GameWorld> GetAll(string environmentId) => GetCached(environmentId);

    public GameWorld? GetByWorldId(string worldId, string environmentId) =>
        GetCached(environmentId).FirstOrDefault(w =>
            string.Equals(w.WorldId, worldId, StringComparison.OrdinalIgnoreCase));

    // ── Single-item cache operations (event-driven) ─────────────

    public async Task UpdateSingleWorldAsync(int id, string environmentId, bool? isVisible = null, CancellationToken ct = default)
    {
        try
        {
            // If explicitly not visible, just remove from cache
            if (isVisible == false)
            {
                Logger.LogInformation("GameWorld isVisible=false, removing from cache (id={Id})", id);
                RemoveFromCache(id, environmentId);
                return;
            }

            // Small delay to ensure backend transaction is committed
            await Task.Delay(100, ct);

            // Fetch single item from API
            var response = await ApiClient.GetAsync<GameWorld>(
                $"/api/v1/server/game-worlds/{id}", ct: ct);

            if (!response.Success || response.Data is null)
            {
                Logger.LogDebug("GameWorld not found (id={Id}), removing from cache", id);
                RemoveFromCache(id, environmentId);
                return;
            }

            UpsertItemInCache(response.Data, environmentId);
            Logger.LogDebug("Single GameWorld upserted in cache (id={Id})", id);
        }
        catch (Exception ex)
        {
            Logger.LogError(ex, "Failed to update single GameWorld (id={Id}), falling back to full refresh", id);
            await FetchByEnvironmentAsync(environmentId, ct);
        }
    }

    public void RemoveWorld(int id, string environmentId)
    {
        RemoveFromCache(id, environmentId);
        Logger.LogInformation("GameWorld removed from cache (id={Id})", id);
    }

    // ── Maintenance helpers ─────────────────────────────────────

    public bool IsWorldMaintenanceActive(string worldId, string environmentId)
    {
        var world = GetByWorldId(worldId, environmentId);
        if (world is null || !world.IsMaintenance) return false;

        var now = DateTime.UtcNow;
        var start = world.MaintenanceStartDate is not null
            ? DateTime.Parse(world.MaintenanceStartDate) : (DateTime?)null;
        var end = world.MaintenanceEndDate is not null
            ? DateTime.Parse(world.MaintenanceEndDate) : (DateTime?)null;

        if (start.HasValue && now < start.Value) return false;
        if (end.HasValue && now > end.Value) return false;
        return true;
    }

    public string? GetWorldMaintenanceMessage(string worldId, string environmentId, string lang = "en")
    {
        var world = GetByWorldId(worldId, environmentId);
        if (world is null) return null;

        if (world.SupportsMultiLanguage == true && world.MaintenanceLocales is { Count: > 0 })
        {
            var locale = world.MaintenanceLocales.FirstOrDefault(l => l.Lang == lang)
                      ?? world.MaintenanceLocales.FirstOrDefault();
            if (locale is not null) return locale.Message;
        }

        return world.MaintenanceMessage;
    }
}
